### PR TITLE
BUG: Preserve subclassing with groupby operations

### LINKED ISSUE #28330: groupby and resample methods do not preserve subclassed data structures
#### Code sample

```python
class MyDataFrame(pd.DataFrame):
    @property
    def _constructor(self):
        return MyDataFrame

dates = pd.date_range('2019', freq='H', periods=1000)
my_df = MyDataFrame(np.arange(len(dates)), index=dates)

print(type(my_df)) 
# __main__.MyDataFrame (✓)

print(type(my_df.diff())) 
# __main__.MyDataFrame (✓)

print(type(my_df.sample(1))) 
# __main__.MyDataFrame (✓)

print(type(my_df.rolling('5H').mean()))
# __main__.MyDataFrame (✓)

print(type(my_df.groupby(my_df.index.dayofweek).mean()))
# pandas.core.frame.DataFrame (✘)

print(type(my_df.resample('D').mean()))
# pandas.core.frame.DataFrame (✘)

```
#### Problem description

Originally posted on [SO](https://stackoverflow.com/questions/57796464/pandas-groupby-resample-etc-for-subclassed-dataframe?noredirect=1#comment102088835_57796464).

The intended behaviour for chain-able methods on subclassed data structures is clearly that the operation returns an instance of the subclass (i.e. `MyDataFrame`), rather than the native type (i.e. `DataFrame`). This is the current behaviour for most operations (e.g., slicing, sampling, sorting) but not resample and groupby.

Currently groupby and resample both return explicitly constructed pandas datatypes, e.g. here:

https://github.com/pandas-dev/pandas/blob/ac693331400bce4747c63aa76c53c6e3488933ae/pandas/core/groupby/generic.py#L338

To get the expected behaviour, the intermediary classes (e.g. DataFrameGroupBy) would need to retain information about the calling class so that the appropriate constructor can be used (i.e. one of `_constructor` or `_constructor_sliced`  or `_constructor_expanddim`).

Note that operations that use Window and Rolling already appear have the expected behaviour because these assemble their results via a call to `concat` such as this one:

https://github.com/pandas-dev/pandas/blob/171c71611886aab8549a8620c5b0071a129ad685/pandas/core/window.py#L325

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.25.1
numpy            : 1.17.1
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.3
setuptools       : 40.8.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.8.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.1
numexpr          : 2.7.0
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : None
tables           : 3.5.2
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by alkasm**
Preserve subclassing with groupby operations

- [x] closes #28330
- [x] tests added / passed
- [x] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -706,7 +708,7 @@ def filter(self, func, dropna=True, *args, **kwargs):
 
         Examples
         --------
-        >>> df = pd.DataFrame({'A' : ['foo', 'bar', 'foo', 'bar',
+        >>> df = pd.self.obj._constructor({'A' : ['foo', 'bar', 'foo', 'bar',
```
Comment (TomAugspurger) [inline pandas/core/groupby/generic.py]: This shouldn't change.

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -134,6 +134,34 @@ def func(dataf):
     result = df.groupby("X", squeeze=False).count()
     assert isinstance(result, DataFrame)
 
+    # https://github.com/pandas-dev/pandas/issues/28330
+    # Test groupby operations on subclassed dataframes/series
+    class ChildSeries(Series):
```
Comment (TomAugspurger) [inline pandas/tests/groupby/test_groupby.py]: We have these in `pandas/util/testing.py` You can reuse them.

File: doc/source/whatsnew/v0.25.2.rst [hunk 3]
```diff
@@ -78,7 +78,7 @@ Groupby/resample/rolling
 
 - Bug incorrectly raising an ``IndexError`` when passing a list of quantiles to :meth:`pandas.core.groupby.DataFrameGroupBy.quantile` (:issue:`28113`).
 - Bug in :meth:`pandas.core.groupby.GroupBy.shift`, :meth:`pandas.core.groupby.GroupBy.bfill` and :meth:`pandas.core.groupby.GroupBy.ffill` where timezone information would be dropped (:issue:`19995`, :issue:`27992`)
--
+- Bug groupby/resampling where subclasses were not returned from groupby ops (:issue:`28330`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.25.2.rst]: Move to 1.0.0.rst.

**Review by TomAugspurger — COMMENTED** [review]
Thanks for the PR. Looks good overall.

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (alkasm) [reply pandas/core/groupby/generic.py]: Whoops

File: pandas/tests/groupby/test_groupby.py (hunk 2 above)
Comment (alkasm) [reply pandas/tests/groupby/test_groupby.py]: Great!

File: doc/source/whatsnew/v0.25.2.rst (hunk 3 above)
Comment (alkasm) [reply doc/source/whatsnew/v0.25.2.rst]: Done.

File: pandas/tests/groupby/test_groupby.py [hunk 4]
```diff
@@ -134,6 +134,23 @@ def func(dataf):
     result = df.groupby("X", squeeze=False).count()
     assert isinstance(result, DataFrame)
 
+    # https://github.com/pandas-dev/pandas/issues/28330
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: ideally we split these into several tests, but could be a followup

File: pandas/tests/groupby/test_groupby.py [hunk 5]
```diff
@@ -134,6 +134,23 @@ def func(dataf):
...
+    # Test groupby operations on subclassed dataframes/series
+
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: actually these tests *are* different that the above ones, can you make a new test

**Review by jreback — CHANGES_REQUESTED** [review]
test change, pls merge master, ping on green.

File: pandas/tests/groupby/test_groupby.py [hunk 6]
```diff
@@ -134,6 +134,23 @@ def func(dataf):
...
+    cdf = tm.SubclassedDataFrame(
+        [
+            {"val1": 1, "val2": 20},
+            {"val1": 1, "val2": 19},
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: also pls parameterize over as many functions that you can here

Comment (WillAyd) [reply pandas/tests/groupby/test_groupby.py]: You can use `reduction_func` from conftest. Probably worth adding something similar for transforming functions as well

**Comment by jbrockmendel** [comment]
@alkasm the fixes here look great.  can you rebase and address comments about making the tests more thorough

File: pandas/tests/groupby/test_groupby.py [hunk 7]
```diff
@@ -134,6 +134,23 @@ def func(dataf):
...
+            {"val1": 2, "val2": 27},
+            {"val1": 2, "val2": 12},
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: can you update to comments here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls also merge master

**Comment by jbrockmendel** [comment]
@alksm can you rebase

**Comment by alkasm** [comment]
Sorry for the delay y'all. Should be able to clean up, write new tests, and rebase this week sometime.

**Comment by alkasm** [comment]
Due to the many changes from #28835 I found it easier to just make a new branch. However there's a few issues I'm running into getting this to work in all cases; and I'm having trouble locating the source of the issues. Here's my current branch off master: https://github.com/alkasm/pandas/tree/alkasm/groupby-preserve-subclass (I can force push the new changes to this current branch, or just open a new PR if it's more helpful)

Borrowing from `tests/groupby/test_transform.py` and the suggestion from @WillAyd  to use the `conftest.reduction_kernels`, I wrote a script to test the transforms and aggregations to see if the subclasses are returned properly:

```python
import pandas as pd
import numpy as np
from pandas.util.testing import SubclassedSeries, SubclassedDataFrame
from pandas.tests.groupby.conftest import reduction_kernels

objs = [
    SubclassedDataFrame(
        dict(a=[0, 0, 0, 1, 1, 1], b=range(6)), index=["A", "B", "C", "D", "E", "F"]
    ),
    SubclassedSeries([0, 0, 0, 1, 1, 1], index=["A", "B", "C", "D", "E", "F"]),
]

for obj in objs:
    g = obj.groupby(np.repeat([0, 1], 3))
    for k in reduction_kernels:
        if k in ("ngroup", "size"):
            continue
        args = {"nth": [0], "quantile": [0.5]}.get(k, [])
        xform = g.transform(k, *args)
        agg = g.agg(k, *args)
        if not (type(obj) == type(xform) == type(agg)):
            print(f"{k:>10}, {obj.__class__.__name__}, {xform.__class__.__name__}, {agg.__class__.__name__}")
```

and from running this there's a few issues:

```
      skew, SubclassedDataFrame, SubclassedDataFrame, DataFrame
   nunique, SubclassedDataFrame, SubclassedDataFrame, DataFrame
       mad, SubclassedDataFrame, SubclassedDataFrame, DataFrame
```

I can't really find how these functions end up constructing their result as a `DataFrame`. Can anyone with some more knowledge here point me in the right direction?

**Comment by jreback** [comment]
can you merge master

**Comment by jbrockmendel** [comment]
Based on a quick look, I think you're best bet is in core.groupby.generic `_wrap_applied_output` methods.

I also recently noticed that core.apply has a place where it does `f(Series([]))` instead of `f(obj._constructor([]))` that might not play well with subclassing.

**Comment by jreback** [comment]
can you merge master

**Comment by WillAyd** [comment]
@alkasm can you fix up merge conflicts?

**Comment by jreback** [comment]
can you merge master and will look again

**Comment by TomAugspurger** [comment]
@alkasm do you have time to fix the merge conflicts? We're doing a release this week.

**Comment by WillAyd** [comment]
Nice PR but looks like this has gone stale. Ping if you'd like to pick back up and can address merge conflicts / comments

**Comment by JBGreisman** [comment]
I recently came across this bug/PR while working on a project that uses a subclassed `DataFrame`. It seems like this PR has been closed and the issue removed from "Contributions Welcome", but the initial issue #28330 is still open. 

I just wanted to check on the current status here -- it seems like this PR was abandoned. If this were to be picked up, would you recommend starting from here or in a fresh PR?

**Comment by WillAyd** [comment]
@JBGreisman if interested you can merge these changes into your own local branch and push up a new PR

**Comment by alkasm** [comment]
@JBGreisman thank you for picking this back up!

**Comment by JBGreisman** [comment]
@alkasm No worries -- thanks for laying the groundwork!

**Pull request closed without merging**